habitat_baselines/common/obs_transformers.py

```diff
@@ -217,113 +218,496 @@ def from_config(cls, config: Config):
+        zxi = xi * d1 + z
+        d2 = torch.sqrt(x2 + y2 + zxi * zxi)
+
+        div = alpha * d2 + (1 - alpha) * zxi
+        u = fx * x / div + cx
+        v = fy * y / div + cy
+
+        # Projected points on image plane
+        # For grid_sample, -1 <= proj_pts <= 1
+        mapx = 2 * u / self.img_w - 1.0
+        mapy = 2 * v / self.img_h - 1.0
+        proj_pts = torch.stack([mapx, mapy], dim=-1)
+
+        # Check valid area
+        if alpha <= 0.5:
+            w1 = alpha / (1 - alpha)
+        else:
+            w1 = (1 - alpha) / alpha
+        w2 = w1 + xi / np.sqrt(2 * w1 * xi + xi * xi + 1)
+        valid_mask = z > -w2 * d1
+        valid_mask *= fov_mask
+
+        return proj_pts, valid_mask
+
+    def unprojection(
+        self, with_rotation: bool = True
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Unpack parameters
+        cx, cy, fx, fy, xi, alpha = self.fish_param
+
+        # Calculate unprojection
+        v, u = torch.meshgrid(
+            [torch.arange(self.img_h), torch.arange(self.img_w)]
+        )
+        mx = (u - cx) / fx
+        my = (v - cy) / fy
+        r2 = mx * mx + my * my
+        mz = (1 - alpha * alpha * r2) / (
+            alpha * torch.sqrt(1 - (2 * alpha - 1) * r2) + 1 - alpha
+        )
+        mz2 = mz * mz
+
+        k1 = mz * xi + torch.sqrt(mz2 + (1 - xi * xi) * r2)
+        k2 = mz2 + r2
+        k = k1 / k2
+
+        # Unprojected unit vectors
+        unproj_pts = k.unsqueeze(-1) * torch.stack([mx, my, mz], dim=-1)
+        unproj_pts[..., 2] -= xi
+
+        # Calculate fov
+        unproj_fov_cos = unproj_pts[..., 2]  # unproj_pts @ z_axis
+        fov_mask = unproj_fov_cos >= self.fov_cos
+        if alpha > 0.5:
+            fov_mask *= r2 <= (1 / (2 * alpha - 1))
+
+        # Rotate unproj_pts points according to camera rotation
+        if with_rotation:
+            unproj_pts = self.camcoord2worldcoord(unproj_pts)
+
+        return unproj_pts, fov_mask
+
+
+class ProjectionConverter(nn.Module):
+    r"""This is the implementation to convert {cubemap, equirec, fisheye} images
+    into {perspective, equirec, fisheye} images.
+    """
+
+    def __init__(
+        self,
+        input_projections: List[CameraProjection],
+        output_projection: CameraProjection,
+    ):
+        """Args:
+        input_projections: (list of CameraProjection) input images of projection models
+        output_projection: (CameraProjection) generated image of projection model
+        """
+        super(ProjectionConverter, self).__init__()
+        self.input_models = input_projections
+        self.output_model = output_projection
+
+        # Check if depth conversion is required
+        # If depth is in z value in input, conversion is required
+        self.input_zfactors = self.calculate_zfactor(self.input_models)
+        # If depth is in z value in output, inverse conversion is required
+        output_invzfactors = self.calculate_zfactor([self.output_model])
```

Comment:
This is already assumes a list of output_projects , so why not just make it support the List[CameraProject] as outputs as well. :)

Reply:
Sure, I will implement `ProjectionConverter` to allow multiple output projections tomorrow (hopefully).